### PR TITLE
Add RegisterEncoder functionality

### PR DESCRIPTION
**Pull request by peter-edge**
For https://github.com/uber-go/zap/issues/330

This is just a pass at this, let me know of any changes obviously. I purposefully didn't do the underscores in front of private variables (I don't want to do this anymore) but I can add this for consistency. We could reuse the global mutex in global.go as well.

### CODE REVIEW

File: encoder.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// If an encoder with the same name already exists, this will return an error.
+// By default, the encoders "json" and "console" are registered.
+func RegisterEncoder(name string, constructor func(zapcore.EncoderConfig) zapcore.Encoder) error {
+	encoderMutex.Lock()
+	defer encoderMutex.Unlock()
+	if name == "" {
+		return errNoEncoderNameSpecified
+	}
+	if _, ok := encoderNameToConstructor[name]; ok {
+		return fmt.Errorf("encoder already registered for name %s", name)
```
Comment (jcorbin) [inline encoder.go]: `%q` may be a moderately more useful formatting code here

File: encoder.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	encoderMutex.RLock()
+	defer encoderMutex.RUnlock()
+	if name == "" {
+		return nil, errNoEncoderNameSpecified
+	}
+	constructor, ok := encoderNameToConstructor[name]
+	if !ok {
+		return nil, fmt.Errorf("no encoder registered for name %s", name)
+	}
+	return constructor(encoderConfig), nil
```
Comment (jcorbin) [inline encoder.go]: Seems a missed opportunity to not let the encoder constructor return an error (e.g. config validation if needed).

File: encoder_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+import (
+	"testing"
+
+	"go.uber.org/zap/zapcore"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var (
```
Comment (jcorbin) [inline encoder_test.go]: is the var binding truly needed?

File: encoder.go (hunk 2 above)
Comment (peter-edge) [inline encoder.go]: Ya I was thinking the same thing, didn't know what to do. I updated to this.

File: encoder_test.go (hunk 3 above)
Comment (peter-edge) [inline encoder_test.go]: Heh no.

File: encoder.go (hunk 1 above)
Comment (peter-edge) [inline encoder.go]: Done

**Comment by peter-edge** [comment]
All comments addressed

**Comment by peter-edge** [comment]
Ya I bet he will too, let's hope he doesn't notice :) Fat chance :)

File: encoder.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap/zapcore"
+)
+
+var (
+	errNoEncoderNameSpecified = errors.New("no encoder name specified ")
+	encoderNameToConstructor  = make(map[string]func(zapcore.EncoderConfig) (zapcore.Encoder, error))
+	encoderMutex              sync.RWMutex
```
Comment (akshayjshah) [inline encoder.go]: nit: please underscore-prefix globals (except for errors).

I'm not wild about this convention either, but let's keep everything consistent.

File: encoder.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+// MustRegisterEncoder registers an encoder constructor for the given name.
+//
+// If an encoder with the same name already exists, this will panic.
+// By default, the encoders "json" and "console" are registered.
+func MustRegisterEncoder(name string, constructor func(zapcore.EncoderConfig) (zapcore.Encoder, error)) {
+	if err := RegisterEncoder(name, constructor); err != nil {
+		panic(err.Error())
+	}
+}
```
Comment (akshayjshah) [inline encoder.go]: Peter, I know you're not wild about this `Must*` pattern; in this case, I'm not a fan either. Can we drop this and encourage users to handle errors properly?

File: encoder.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		return nil, fmt.Errorf("no encoder registered for name %q", name)
+	}
+	return constructor(encoderConfig)
+}
+
+func wrapEncoderConstructorNoError(constructor func(zapcore.EncoderConfig) zapcore.Encoder) func(zapcore.EncoderConfig) (zapcore.Encoder, error) {
+	return func(encoderConfig zapcore.EncoderConfig) (zapcore.Encoder, error) {
+		return constructor(encoderConfig), nil
+	}
+}
```
Comment (akshayjshah) [inline encoder.go]: Could we simplify a bit by declaring `encoderNameToConstructor` as a map literal (or the output of a `defaultEncoders()` function)? I think that'll let us drop this wrapper function, `MustRegisterEncoder`, `registerDefaultEncoders`, and `init`.

**Comment by akshayjshah** [comment]
Haha, didn't see the discussion about underscores 😄

File: encoder.go (hunk 6 above)
Comment (peter-edge) [inline encoder.go]: Done

File: encoder.go (hunk 5 above)
Comment (peter-edge) [inline encoder.go]: Done

File: encoder.go (hunk 4 above)
Comment (peter-edge) [inline encoder.go]: Done

**Comment by peter-edge** [comment]
All comments addressed

**Comment by akshayjshah** [comment]
Last nit: please add messages in assertions.

File: encoder.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		return nil, errNoEncoderNameSpecified
+	}
+	constructor, ok := _encoderNameToConstructor[name]
+	if !ok {
+		return nil, fmt.Errorf("no encoder registered for name %q", name)
+	}
+	return constructor(encoderConfig)
+}
+
+func defaultEncoders() map[string]func(zapcore.EncoderConfig) (zapcore.Encoder, error) {
```
Comment (jcorbin) [inline encoder.go]: Do we even need this function now?

**Comment by peter-edge** [comment]
Both comments should be addressed

**Comment by peter-edge** [comment]
@akshayjshah note you have to merge this :)

**Pull request merged**